RecoMET/METAlgorithms/src/CSCHaloAlgo.cc

```diff
@@ -545,14 +594,67 @@ reco::CSCHaloData CSCHaloAlgo::Calculate(const CSCGeometry& TheCSCGeometry,
        
        float iPhi = iGlobalPosition.phi();
        float iR =  TMath::Sqrt(iGlobalPosition.x()*iGlobalPosition.x() + iGlobalPosition.y()*iGlobalPosition.y());
-       short int nSegs = 0;
+       float iZ = iGlobalPosition.z();
+       float iT = iSegment->time();
+       //       if(fabs(iZ)<650&& TheEvent.id().run()< 251737) iT-= 25; 
+       //Calo matching:
+       reco::Vertex::Point vtx(0,0,0);
+       for(size_t ihit = 0; ihit<hbhehits->size(); ++ ihit){
+	 if(calomatched) break;
+	 const HBHERecHit & rechit = (*hbhehits)[ ihit ];
+	 math::XYZPoint rhpos = getPosition(rechit.id(),vtx);
+	 double rhet = rechit.energy()/cosh(rhpos.eta()); 
+	 double dphi_rhseg = fabs(deltaPhi(rhpos.phi(),iGlobalPosition.phi()));
+	 double dr_rhseg = sqrt(rhpos.x()*rhpos.x()+rhpos.y()*rhpos.y()) - iR;
+	 double dtcorr_rhseg = rechit.time()- fabs(rhpos.z()-iZ)/30- iT; // 
+	 if(rhet>10&& 
+	    dphi_rhseg <0.05 && 
+	    dr_rhseg< 25 && dr_rhseg> -25 && //careful: asymmetric cut might not be the most appropriate thing 
+	    dtcorr_rhseg>0    
+	    ) calomatched= true;
+       }
 
+       for(size_t ihit = 0; ihit<ecalebhits->size(); ++ ihit){
+	 if(calomatched) break;
+	 const EcalRecHit & rechit = (*ecalebhits)[ ihit ];
+	 math::XYZPoint rhpos = getPosition(rechit.id(),vtx);
+	 double rhet = rechit.energy()/cosh(rhpos.eta()); 
+	 double dphi_rhseg = fabs(deltaPhi(rhpos.phi(),iGlobalPosition.phi()));
+	 double dr_rhseg = sqrt(rhpos.x()*rhpos.x()+rhpos.y()*rhpos.y()) - iR;
+	 double dtcorr_rhseg = rechit.time()- fabs(rhpos.z()-iZ)/30- iT; // 
+	 if(rhet>10&& 
+	    dphi_rhseg <0.05 && 
+	    dr_rhseg< 25 && dr_rhseg> -25 && //careful: asymmetric cut might not be the most appropriate thing 
+	    dtcorr_rhseg>0    
+	    ) calomatched= true;
+       }
+       for(size_t ihit = 0; ihit<ecaleehits->size(); ++ ihit){
+         if(calomatched) break;
+         const EcalRecHit & rechit = (*ecaleehits)[ ihit ];
+	 math::XYZPoint rhpos = getPosition(rechit.id(),vtx);
+         double rhet = rechit.energy()/cosh(rhpos.eta());
+         double dphi_rhseg = fabs(deltaPhi(rhpos.phi(),iGlobalPosition.phi()));
+         double dr_rhseg = sqrt(rhpos.x()*rhpos.x()+rhpos.y()*rhpos.y()) - iR;
+         double dtcorr_rhseg = rechit.time()- fabs(rhpos.z()-iZ)/30- iT; //                                                                       
+         if(rhet>10&&
+            dphi_rhseg <0.05 &&
+            dr_rhseg< 25 && dr_rhseg> -25 && //careful: asymmetric cut might not be the most appropriate thing                                  
+	    dtcorr_rhseg>0
+            ) calomatched= true;
+     }
```

Comment:
If the three "for" loops in lines 602-644 are intended to be the same, they should be made into a function that is called three times, once each for hbhehits, ecalebhits, and ecaleehits. Maybe there is some subtle difference I'm missing. But copy-and-paste programming should be avoided.